docs/dev-client-gameplay.md

```diff
@@ -253,475 +454,702 @@ const canBeSpawnPoint = Waypoint.flags[eid] & WaypointFlags.canBeSpawnPoint;
-}
+import { addEntity } from "bitecs";
+import { BoxGeometry, Mesh, MeshBasicMaterial } from "three";
+import { addObject3DComponent } from "../src/utils/jsx-entity";
+
+const eid = addEntity(world);
+addObject3DComponent(world, eid, new Mesh(
+  new BoxGeometry(1.0, 1.0, 1.0),
+  new MeshBasicMaterial({ color: 0xffffff })
+));
 ```
 
-Notice that `GrabbableParams` do not map one-to-one with runtime components. That is, there is no `Grabbable` component. This is common in situations where we want to expose user-friendly options (for use in Blender, Spoke, or when writing `EntityDef` s), while representing the information differently at runtime.
+`addObject3DComponent()` checks whether an `Object3D` is associated for an
+entity. It throws an exception if already asssociated. It prevents multiple
+`Object3D` associations.
 
+```typescript
+const eid = addEntity(world);
+addObject3DComponent(world, eid, new Group());
+// Throws an exception
+addObject3DComponent(world, eid, new Group());
+```
 
-<a id="orga810c47"></a>
+`addObject3DComponent()` adds the built-in component `Object3DTag` defined in
+[`src/bit-components.js`](https://github.com/mozilla/hubs/blob/master/src/bit-components.js)
+to an entity and stores an `Object3D` in the special `eid2obj` map managed in
+`HubsWorld`.
 
-### Default inflators
+`bitECS` component doesn't natively support non-numerical data type. So we
+decided to store `Object3D`s associated with entities in a special map
+`HubsWorld.eid2obj` that maps from Entity ID to `Object3D`.
 
-For many components, the `EntityDef` or `GLTF` representation DOES match the runtime component, and writing individual inflators for each would be unnecessary boilerplate.
+TODO: Replace `HubsWorld.eid2obj` with `Object3DTagMap` to be more aligned with the pattern
+described above?
 
-For example, the `SceneLoader` component has a single property, a string `src`:
+You can use `Object3DTag` component with query to find entities
+that have associated `Object3D`, and access the associated `Object3D`s with
+`HubsWorld.eid2obj` map.
 
 ```typescript
-export const SceneLoader = defineComponent({ src: Types.ui32 });
-SceneLoader.src[$isStringType] = true;
+const objectQuery = defineQuery([Object3DTag]);
+export function object3DSystem(world: HubsWorld) {
+  objectQuery(world).forEach(eid => {
+    const obj = world.eid2obj.get(eid)!;
+    ...
+  });
+}
 ```
 
-It can be specified in the following `EntityDef` :
+`Object3DTag` component and `HubsWorld.eid2map` manipulations are hidden in
+the Hubs Client core. They should be read-only from add-ons.
 
-```typescript
-export function ScenePrefab(src: string): EntityDef {
-  return <entity name="Scene" sceneRoot sceneLoader={{ src }} />;
-}
-```
 
-Writing the inflator for the `SceneLoader` component is simple, except for the fact that we have to remember to convert the `string` to a `StringID`, so that the `bitECS` component can hold onto it:
+### EntityDef, JSX, Prefab

```

Comment:
Title the section after the desired outcome “Adding an entity”? Then you could break it down into separate steps for defining an entity and adding an entity to the world.

Reply:
Thanks for the suggestion. I have been thinking about it but don't have a conclution yet. Let me merge this PR as is for now, and let's think of making a follow up PR if needed.